Add pagination support for commit ancestors

indradhanush opened the pull request:
Updates GraphQL API for https://github.com/sourcegraph/sourcegraph/issues/44993. 

## Notes on approach

### Update 2022-12-08
After reviews and suggestions, approach has been updated. See [comment
](https://github.com/sourcegraph/sourcegraph/pull/45157#issuecomment-1343098557)

The section below is outdated but offers insight into the thought process that went into this PR early on.

Here are the following approaches (in order) that I tried and reverted at various stages before settling on the final solution proposed in this PR:

1. Add an `after` arg based on pagination guidelines from Relay [GraphQL Cursor Connections Specification](https://relay.dev/graphql/connections.htm)
	1. There is already a argument on the `ancestors` object named `after` which corresponds to the `--after` flag in the `git rev-list` / `git log` command and is used to indicate commits after a specific timestamp
2. Considered adding `afterCommit` with the following idea in mind:
	1. Deprecate `after` and add `afterTImestamp` to take its place
	2. Remove `after`
	3. Deprecate `afterCommit` add add `after` which would now imply to indicate after which commit the API should return the results

This approach sounded too long and complicated and I finally considered adding a field `revision` to the `ancestors` fragment.

But the cleanest way of all right now is to rely on the `GitCommit.Range` field since it already accepts a `rev` argument to indicate which commit should be returned by the API. In combination with the `first` argument to the `ancestors` field which is already in use.

The problem to be solved is simplified to only return a `cursor` in the `pageInfo` object so that clients can determine the next value for `rev` in subsequent API calls. 

In this PR, the `pageInfo` object will now contain an `endCursor` which is in accordance to the recommended pagination guidelines in the GraphQL spec. The only divergence here is not using an argument named `after` or `offset` in the `ancestors` field - a divergence which is well justified instead of trying to stick to the convention in exchange for a complicated API.

The final step before a client can use the value in the `endCursor` field is that they will need to decode the base64 string to reteive the commit spec - which is in the same format as that of a `GitCommit.ID` in our API that is a GraphQL ID.

## Test plan

1. Builds pass
2. Added tests
3. Tested locally

Example API call to test locally can be seen by following [this](https://sourcegraph.test:3443/api/console?trace=1#%7B%22query%22%3A%22query%20FetchCommits(%24repo%3A%20ID!%2C%20%24revision%3A%20String!%2C%20%24first%3A%20Int%2C%20%24currentPath%3A%20String%2C%20%24query%3A%20String%2C%20%24afterCursor%3A%20String)%20%7B%5Cn%20%20node(id%3A%20%24repo)%20%7B%5Cn%20%20%20%20__typename%5Cn%20%20%20%20...%20on%20Repository%20%7B%5Cn%20%20%20%20%20%20commit(rev%3A%20%24revision)%20%7B%5Cn%20%20%20%20%20%20%20%20ancestors(first%3A%20%24first%2C%20query%3A%20%24query%2C%20path%3A%20%24currentPath%2C%20afterCursor%3A%20%24afterCursor)%20%7B%5Cn%20%20%20%20%20%20%20%20%20%20...CommitAncestorsConnectionFields%5Cn%20%20%20%20%20%20%20%20%7D%5Cn%20%20%20%20%20%20%7D%5Cn%20%20%20%20%7D%5Cn%20%20%7D%5Cn%7D%5Cn%5Cnfragment%20CommitAncestorsConnectionFields%20on%20GitCommitConnection%20%7B%5Cn%20%20nodes%20%7B%5Cn%20%20%20%20...GitCommitFields%5Cn%20%20%7D%5Cn%20%20pageInfo%20%7B%5Cn%20%20%20%20hasNextPage%5Cn%20%20%20%20endCursor%5Cn%20%20%7D%5Cn%7D%5Cn%5Cnfragment%20GitCommitFields%20on%20GitCommit%20%7B%5Cn%20%20abbreviatedOID%5Cn%20%20subject%5Cn%20%20author%20%7B%5Cn%20%20%20%20person%20%7B%5Cn%20%20%20%20%20%20name%5Cn%20%20%20%20%7D%5Cn%20%20%7D%5Cn%7D%22%2C%22operationName%22%3A%22FetchCommits%22%2C%22variables%22%3A%22%7B%5Cn%20%20%5C%22currentPath%5C%22%3A%20%5C%22bill-of-materials.json%5C%22%2C%5Cn%20%20%5C%22first%5C%22%3A%205%2C%5Cn%20%20%5C%22afterCursor%5C%22%3A%20%5C%222%5C%22%2C%5Cn%20%20%5C%22query%5C%22%3A%20%5C%22%5C%22%2C%5Cn%20%20%5C%22repo%5C%22%3A%20%5C%22UmVwb3NpdG9yeToz%5C%22%2C%5Cn%20%20%5C%22revision%5C%22%3A%20%5C%2209304a4%5C%22%5Cn%7D%22%7D) link provided your local dev instance already has the `github.com/sourcegraph-testing/etcd` repo cloned and available to serve search requests. 

You will probably need to update the value of `repo` in the variables to reflect the ID from your database.



<!-- All pull requests REQUIRE a test plan: https://docs.sourcegraph.com/dev/background-information/testing_principles -->

mrnugget commented on cmd/frontend/graphqlbackend/git_commit_test.go:
```diff
@@ -180,3 +181,183 @@ func TestGitCommitFileNames(t *testing.T) {
+	}
+	c2 := gitdomain.Commit{
+		ID: api.CommitID("ccdde12345"),
+		Parents: []api.CommitID{
+			c1.ID,
+		},
```
```suggestion
		Parents: []api.CommitID{c1.ID},
```

mrnugget commented on cmd/frontend/graphqlbackend/git_commit_test.go:
```diff
@@ -180,3 +181,183 @@ func TestGitCommitFileNames(t *testing.T) {
+
+	// Start at commit c1.
+	// Expect c1 and c2 in the nodes. c3 in the endCursor.
+	RunTest(t, &Test{
```
Why not use `RunTests` (plural) if you have multiple tests?

mrnugget reviewed (commented):
Hmmm, I'm not fully onboard with this approach. It adds an `endCursor` that's not really an opaque cursor, because you can't use it when querying, you need to decode it first.

I think a solution would ideally provide and make the user of a proper opaque cursor possible.

Maybe we can add an `afterCursor: string` field to `ancestors` that then takes in that cursor? And deprecate `after` and slowly switch to a `afterDate: DateTime` field instead?

indradhanush commented:
> Hmmm, I'm not fully onboard with this approach. It adds an endCursor that's not really an opaque cursor, because you can't use it when querying, you need to decode it first.
>
>
>
>I think a solution would ideally provide and make the user of a proper opaque cursor possible.
>
>Maybe we can add an afterCursor: string field to ancestors that then takes in that cursor? And deprecate after and slowly switch to a afterDate: DateTime field instead?

Agreed. I was thinking about this and my thoughts is that `rev` should be deprecated and we should replace that with something like `revID` so that the cursor can be used opaquely instead of the other way round. What do you think?

mrnugget commented:
> Agreed. I was thinking about this and my thoughts is that `rev` should be deprecated and we should replace that with something like `revID` so that the cursor can be used opaquely instead of the other way round. What do you think?

I think `rev` is fine. Since it's the entry point for a list. Technically we could just use `rev` as the pointer. Client would need to fetch n+1 commits and then using the additonal commit as the starting point when requesting the next page. Right?

Take a look at other paginated connections in our API to see how we use `after`.

indradhanush commented:
> I think rev is fine. Since it's the entry point for a list. Technically we could just use rev as the pointer. Client would need to fetch n+1 commits and then using the additonal commit as the starting point when requesting the next page. Right?

Since we're already using `rev` to fetch the n+1 commits isn't it kind of redundant to have to set the proposed `afterCursor` in the `ancestors` again?

indradhanush commented:
Also, for a second request, the client will need to decode the `endCursor` anyway since the `rev` argument for `commit` currently accepts a revSpec-ish string.

indradhanush commented:
@philipp-spiess Requested a review from you since you're also looking at this.

philipp-spiess commented:
Hey! Hehe I looked into the [same thing today](https://github.com/sourcegraph/sourcegraph/issues/43658#issuecomment-1337601381) (good thing I reached out to @mrnugget about my ideas 😅). 

Some thoughts:

- I was initially coming from the direction of using the [new pagination abstractions](https://github.com/sourcegraph/sourcegraph/pull/45167) that we're working on but I noticed quickly that git won't allow us to do a backward pagination (so we can't effectively implement `last:` 😭).
- I also noticed that we already define an `after` value for this query that means something different than the `after` value of other queries 😭 `after` is supposed to get the opaque token returned form the `pageInfo` response and I agree that we can't easily delete fields form GraphQL).
- My proposal would do the following:
  - Include the `afterCursor` in the `pageInfo` that is the SHA of the last visited commit (we can use the marshalled ID here). A cursor needs to uniquely identify the item and the sha is already doing.
  - For fetching the next page, the client will the pass this field as-is as the new `afterCursor`.
  - On the backend, if an afterCursor is set, we need to increase the limit by one and skip the first result to avoid double-rendering the commit. The `afterCursor` can be decoded on the backend and used as the new <range>
  
  I think this is pretty much what you had in mind as well but making the server smarter and the client dumb? 🙈

mrnugget commented:
> * My proposal would do the following:

Agree with that! That's what I meant in the last paragraph here: https://github.com/sourcegraph/sourcegraph/pull/45157#pullrequestreview-1204519136

felixfbecker commented:
How does this implementation handle interleaved diverging histories (merge commits)? See https://github.com/sourcegraph/sourcegraph/issues/43658#issuecomment-1338132880

philipp-spiess commented:
So if I understand the implementation correctly, this uses [`git log`](https://sourcegraph.sourcegraph.com/github.com/sourcegraph/sourcegraph@ps/blame-recency/-/blob/internal/gitserver/commands.go?L1747-1762) under the hood and changes the _range_.

So I reproduced this history locally and... Yeah 😢 

```bash
α playground (main) git --no-pager log --pretty=oneline
44544c4c338ba7325551e6502e29723fea1b3d9f (HEAD -> main) main commit 5
ff9d7e65f793f5120497ef4c84a0ad3b8bfe748e Merge branch 'branch'
394652ae02c70ed3488ab8f97306c9cb5cd0daec main commit 4
d4b1827c474c810172a04939797d6360805c46ca (branch) branch commit 3
40f1c30cefbfcaeb3efb35c0d9a28f21d5915e71 main commit 3
d69abd683ccdf9906e1837e4e327f73893ab328e branch commit 2
8e4129971b917e5b40fcaafcc8378916b1a9cefe main commit 2
457e1ef9f3e6e455d2e07797fec9cbea7ac53a08 branch commit 1
43f44437d3476ff2029f2ba4fe4e2cc683cd2c11 main commit 1
α playground (main) git --no-pager log --pretty=oneline d4b1827c474c810172a04939797d6360805c46ca
d4b1827c474c810172a04939797d6360805c46ca (branch) branch commit 3
d69abd683ccdf9906e1837e4e327f73893ab328e branch commit 2
457e1ef9f3e6e455d2e07797fec9cbea7ac53a08 branch commit 1
43f44437d3476ff2029f2ba4fe4e2cc683cd2c11 main commit 1
α playground (main)
```

I don't think we can use the range here in this case. :( 

What we could still do is:

- Use `--skip` to simulate an offset based pagination
- Use `--before` and the timestamp. For this, we'll have to take care that we do not skip commits that share the same timestamp so we would need to do something like this:
   - Get a timestamp that is `1sec` before the one of the commit that we want to show items after.
   - Increase `-n` in some way (2x?)
   - Find the last visible commit in the result set:
      - If found: 
         - If there are _n_ commits afterwards, serve them
         - Otherwise increase `n` so that this will be the case.
      - If not found:
         - Increase `n` and repeat
  
  So this is not going to be an easy road and it's unclear if it brings value over `--skip` (does git maintain an index on the timestamp?). So I'd say we go with `--skip` and encode the offsets in the pagination cursor :/ 

  ```bash
  α playground (main) git --no-pager log --pretty=oneline --before="2022-12-06 10:37:44"
  d4b1827c474c810172a04939797d6360805c46ca (branch) branch commit 3
  40f1c30cefbfcaeb3efb35c0d9a28f21d5915e71 main commit 3
  d69abd683ccdf9906e1837e4e327f73893ab328e branch commit 2
  8e4129971b917e5b40fcaafcc8378916b1a9cefe main commit 2
  457e1ef9f3e6e455d2e07797fec9cbea7ac53a08 branch commit 1
  43f44437d3476ff2029f2ba4fe4e2cc683cd2c11 main commit 1
  ```

felixfbecker commented on cmd/frontend/graphqlbackend/schema.graphql:
```diff
@@ -4545,6 +4545,11 @@ type GitCommit implements Node {
+        """
+        Skip the first N commits of the repo before returning the number of commits as set in the field "first".
+        """
+        offset: Int
```
I agree with @mrnugget that it would be better to introduce a parameter like `afterCursor`, an alias for `after` like `afterDate`, and deprecating the current `after` so that eventually we can change it to be compliant with Relay and remove this footgun vs doubling down on it by introducing other pagination params that are not opaque

indradhanush replied:
@felixfbecker I've renamed `offset` -> `afterCursor`. I will follow up with deprecating `after` in another PR as it was changing more lines than I'd like to have in this PR. 

How does that sound to you?

re: Opaque cursor

At the moment it's an int with the offset that the client can opaquely use in the subsequent request. Is this not what you're thinking? Do I need to change something here?

indradhanush replied:
`afterCursor` should also be a string type right? That's preventing it from being completely opaque at the moment. 🤔

mrnugget commented on cmd/frontend/graphqlbackend/git_commit_test.go:
```diff
@@ -180,3 +180,220 @@ func TestGitCommitFileNames(t *testing.T) {
 }
+
+func TestGitCommitAncestors(t *testing.T) {
+	externalServices := database.NewMockExternalServiceStore()
+	externalServices.ListFunc.SetDefaultReturn(nil, nil)
```
Do you need this?

indradhanush replied:
I didn't. Removed.

mrnugget commented on cmd/frontend/graphqlbackend/git_commits.go:
```diff
@@ -30,29 +37,52 @@ type gitCommitConnectionResolver struct {
+		// same as not setting the flag.
+		afterCursor, err := r.afterCursorAsInt()
+		if err != nil {
+
```
Should return err

indradhanush replied:
[cbafef6](https://github.com/sourcegraph/sourcegraph/pull/45157/commits/cbafef6c667fef0e10dbf9214eba409a8cbc70c4)

mrnugget commented on cmd/frontend/graphqlbackend/git_commits.go:
```diff
@@ -101,7 +131,36 @@ func (r *gitCommitConnectionResolver) PageInfo(ctx context.Context) (*graphqluti
-	return graphqlutil.HasNextPage(r.first != nil && len(commits) > 0 && len(commits) > int(*r.first)), nil
+	totalCommits := len(commits)
+	// If no limit is set, we have retrieved all the commits and there is no next page.
+	if r.first == nil {
```
What if `*r.first == 0`?

indradhanush replied:
If `*r.first == 0`, then we will fetch `1` commit (we always fetch `first + 1` to determine `hasNextPage`). 

The check after this for `totalCommits > limit` will still pass and although 0 commits will be returned, `hasNextPage` will be set to `true`. And if `afterCursor` is also used along with `first: 0` such that no commits exist beyond the `afterCursor`, then `hasNextPage` will be `false` and no commits will be returned.

indradhanush replied:
I've added tests to validate this.

mrnugget replied:
Thanks!

indradhanush commented:
> LGTM. Did you confirm that it works locally in your instance by using API explorer?

Yes, I did.

I added some screenshots below for visual review:

**first: 5, afterCursor: ""**

`endCursor` is `5`.

<img width="1974" alt="image" src="https://user-images.githubusercontent.com/2682729/206962034-4276b665-8a21-4592-9a03-eb8c80debb32.png">

**first: 5, afterCursor: "2"**

Commits 1, 2 and 3 are the same commits from the above screenshot indexed 3, 4 and 5 respectively.

`endCursor` has also updated accordingly.

<img width="1975" alt="image" src="https://user-images.githubusercontent.com/2682729/206962148-13f2daff-01be-4186-bbe6-adc080240b9a.png">
